packages/runtime/container-runtime/src/test/gc/gcBenchmarks.spec.ts

```diff
@@ -0,0 +1,207 @@
+				createContainerRuntimeVersion: pkgVersion,
+				createContainerTimestamp: Date.now(),
+			},
+			isSummarizerClient,
+			readAndParseBlob: async <T>(id: string) => gcBlobsMap.get(id) as T,
+			getNodePackagePath: async (nodeId: string) => testPkgPath,
+			getLastSummaryTimestampMs: () => Date.now(),
+			submitMessage: (message: ContainerRuntimeGCMessage) => {},
+			sessionExpiryTimerStarted: createParams.sessionExpiryTimerStarted,
+		}) as GcWithPrivates;
+	}
+
+	beforeEach(() => {
+		mockLogger = new MockLogger();
+		mc = mixinMonitoringContext(mockLogger, configProvider);
+		garbageCollector = createGarbageCollector({
+			createParams: {
+				gcOptions: { inactiveTimeoutMs },
+			},
+		});
+	});
+
+	afterEach(() => {
+		mockLogger.clear();
+		configProvider.clear();
+		defaultGCData = { gcNodes: {} };
+		garbageCollector?.dispose();
+	});
+
+	/**
+	 * These tests benchmark how long it takes for the initializeOrUpdateGCState function to run
+	 * when there are a high number of unreferenced nodes. Since this results in clearing and
+	 * setting of timers, it takes a long time and can cause slow downs.
+	 */
+	// The benchmark runs for the following number of unreferenced nodes.
+	const unrefNodeCounts = [5000, 15000, 30000];
+	for (const unrefNodeCount of unrefNodeCounts) {
+		it(`initializeOrUpdateGCState with ${unrefNodeCount} nodes`, async () => {
+			const currentTime = Date.now();
+			// Set the unreferenced timestamp to older than inactive timeout so that the nodes start
+			// as inactive.
+			const unreferencedTimestampMs = currentTime - (inactiveTimeoutMs + 10);
+
+			// Generate a base snapshot that has `unrefNodeCount` number of unreferenced nodes
+			// and all are inactive.
+			const gcState: IGarbageCollectionState = { gcNodes: {} };
+			for (let i = 0; i < unrefNodeCount; i++) {
+				gcState.gcNodes[`node${i}`] = {
+					outboundRoutes: [],
+					unreferencedTimestampMs,
+				};
+			}
+			const baseSnapshot: IGarbageCollectionSnapshotData = {
+				gcState,
+				tombstones: [],
+				deletedNodes: [],
+			};
+
+			// Override the base snapshot promise to return the above snapshot.
+			garbageCollector.baseSnapshotDataP = new LazyPromise<
+				IGarbageCollectionSnapshotData | undefined
+			>(async () => baseSnapshot);
+
+			// Call initializeOrUpdateGCState once so that it initializes the GC state from the
+			// base snapshot. The calls to initializeOrUpdateGCState after this one happens on
+			// every connection and that's the one we are benchmarking
+			await garbageCollector.initializeOrUpdateGCState();
+
+			// Repeat the benchmark at least twice and get the average.
+			const repeatCount = 2;
+			let totalTime: number = 0;
+			for (let i = 0; i < repeatCount; i++) {
+				const s = performance.now();
+				await garbageCollector.initializeOrUpdateGCState();
+				const e = performance.now();
+				totalTime += e - s;
+			}
+			console.log(`Avg time for ${unrefNodeCount} nodes: ${totalTime / repeatCount}`);
```

Comment:
If nothing else, maybe just check in the results you see and then we can run it later and compare?  Obviously whatever we can do to get the results on consistent hardware is good, maybe we can get some output in the pipeline but all my ideas there are really hacky.